[WIP] KokkosKernels: migrating sparse m-m addition from Tpetra

brian-kelley opened the pull request:
-Separated symbolic and numeric parts of kernel
-Created SPADDHandle
-Kernels take rowmap/entries/values arrays instead of KokkosSparse CrsMatrix
-"input sorted" information passed through handle
-replaced unsorted symbolic algorithm (see comments in KokkosSparse_spadd.hpp for details)
-Improved unit tests and migrated them into KokkosKernels (testing on the 4 device types)

mhoemmen commented:
@brian-kelley Just curious -- what happens if `HAVE_KOKKOSKERNELS_EXPERIMENTAL` is not defined?  Does it go to Tpetra's fall-back sequential host implementation?

brian-kelley commented:
@mhoemmen Yes. It would be nice though to always have the kernels enabled.

mhoemmen commented:
@brian-kelley I would very much like the kernels to be built and tested by default.  I think it would be good to make that a separate commit / PR, though, in order to facilitate bisection.  Thanks!

srajama1 commented on packages/kokkos-kernels/src/common/KokkosKernels_Handle.hpp:
```diff
@@ -331,6 +337,22 @@ class KokkosKernelsHandle{
+    return this->spaddHandle;
+  }
+
+  void create_spadd_handle(bool input_sorted) {
```
I thought this method is idempotent. Instead it deletes old handles which means the user has the responsibilty to call it right. It is ok for this commit, as it follows spgemm style, but can you file an issue to remove this delete()  or to describe the case for it clearly.

mndevec replied:
This is the way I followed overall. We can talk on the meeting and I can explanation more clearly

srajama1 replied:
understood. Brain doesn't have to be blocked by this.

srajama1 commented on packages/kokkos-kernels/src/sparse/KokkosSparse_spadd.hpp:
```diff
@@ -0,0 +1,681 @@
+
+  /*
+    Unsorted symbolic algorithm notes:
+      -Only needs to sort and merge indices once, in symbolic (sorting is expensive)
```
I think something wrong with tabs

brian-kelley replied:
Fixed in new commits

srajama1 commented on packages/kokkos-kernels/src/sparse/KokkosSparse_spadd.hpp:
```diff
@@ -0,0 +1,681 @@
+      -See UnsortedNumericSumFunctor below for the usage of Apos/Bpos
+  */
+
+  //get C rowmap for sorted input
```
Trying to mix types between A and B here ? Why are they separate parameters ? Can we have default where types are  same.

brian-kelley replied:
@srajama1 I can do that. Mehmet suggested by email to have the A/B/C types all specified separately, but I think that was just because he copy-pasted it the function prototypes from spgemm. There will never be a use case for that in Tpetra, and it would make the code look a lot cleaner.

mndevec replied:
Siva, Brian, 
It is not the "data type" (whether it is float/double integer) that is mixed through with this interface. 
Having different types for A/B/C allows us to pass views with different memory spaces. 
We make sure that the data types are same within within our specialization layer.

srajama1 replied:
I got that after typing that comment ! I am not sure we need this functionality for spadd though. It is just complexity we would have to maintain for a long time.

mndevec replied:
Actually, as long as the main interface allows different types for A/B/C we do not have to propagate all this types within the code to the functors. We will do the type casting in the specialization layer,  then functors will not need to worry about this. I can help with this.

brian-kelley replied:
@mndevec Thanks for explaining! My understanding is that all the internal functors should have template parameters scalar_t, size_type, index_type, A_memory_space, B_memory_space, C_memory_space?

And then in add_symbolic/add_numeric we keep the current template parameters, but static_assert that A, B and C scalar/size_type/index_type all match?

mndevec replied:
@brian-kelley 
Firstly, you do not need to worry most of these things. The current way looks good to me. If you push these changes to kokkos-kernels, I can take care of the rest, and point you to these changes. But below are my answers.
 
Yes, for the second question. 

The general interface is as general as possible. You can see the example here:  https://github.com/kokkos/kokkos-kernels/blob/develop/src/sparse/KokkosSparse_spgemm_numeric.hpp

We want this interface to be callable by users without requiring them to change/cast their views. 
The only requirement we have is that "data types should match". This interface will make sure that data types matches for different variants. 

Within the SpGEMM code, there is no operation that depends on the memory space of these views. I think this is almost true for all codes we have there, unless there is some extra specialization one does. So, this interface will cast all memory spaces to unmanaged memory spaces starting from line 167, so that, we do not compile again and again for different memory spaces. 

Than at line 249, I call the internal function. At this point, with casts, you limit the input types, therefore you are free to choose how the internal functions are restricted. For the first part of your question, you can do it freely as you like, as after this entry point you are limiting the input types. 

The internal code I have for spgemm still uses all different templates, as the entry point is implemented after all is implemented in my case :). But still, I use all templates in the internal code, because it allows me to maximize the code reuse internally. 

By the way, I don't know if distance-2 coloring is included in this push. But you do not need to worry about that either. We started integrating your distance-2 coloring code as practice with Will (https://github.com/william76/kokkos-kernels/blob/distance-2-coloring/src/graph/KokkosGraph_D2_GraphColor.hpp). So, you do not need to spend time on that.

srajama1 commented:
@brian-kelley : Some small comments. BTW, the kokkoskernels portion of this should be generated as a patch and applied to https://github.com/kokkos/kokkos-kernels develop branch as the next synch from there to Trilinos will overwrite this. The Trilinos version is just a snapshot.

mndevec commented:
Brian, sorry for the more work. If you push the kokkos-kernels changes to  https://github.com/kokkos/kokkos-kernels ; that would be much easier. 
We will need the tpetra changes during the kokkos-kernels integration.

mndevec commented:
@mhoemmen ,
By the way, we do not have any code that is surrounded with HAVE_KOKKOSKERNELS_EXPERIMENTAL flag in kokkos kernels. All of the kokkos-kernels codes are compiled and enabled with or without this flag enabled. I think HAVE_KOKKOSKERNELS_EXPERIMENTAL is only used for the integration of SpGEMM right now, because of some MueLu test failures. 

The use of the flag is currently for safer integration. In this case, I think Brian can remove the flag as his code was already used default in Tpetra.

mhoemmen commented:
@mndevec Thanks for clarifying!

brian-kelley commented:
@mndevec @srajama1 @mhoemmen I made a pull request in kokkos-kernels: https://github.com/kokkos/kokkos-kernels/pull/122

I fixed some general type name consistency and size_type vs. ordinal_type issues, and added static_asserts in symbolic/numeric to check that A/B/C data types match, and that C data types are nonconst. I didn't change the functor template arguments or cast the input views to a common unmanaged memory space like SPGEMM does - I'll leave that to @mndevec.

Since there were changes to KokkosKernels_Handle.hpp since the last kokkos-kernels develop snapshot was copied to Trilinos develop, there are some small differences (typedef names, slightly different create/destroy_spadd_handle()) in this pull request and the one in kokkos-kernels. When the next snapshot happens, I can make the small changes to Tpetra add() that uses the kernels.

trilinos-autotester commented:
Status Flag 'Pre-Test Inspection' - Auto Inspected - Inspection Is Not Necessary for this Pull Request.

trilinos-autotester commented:
Status Flag 'Pull Request AutoTester' - Testing Jenkins Projects:

Pull Request Auto Testing STARTING
------------------------------------

## Build Information
- Test Name: Trilinos_autotester_test
- Build Num: 125
- Status: STARTED

### Jenkins Parameters
| Parameter Name | Value |
|----------------|-------|
| COMPILER_MODULE | sems-gcc/4.8.4 |
| JENKINS_BUILD_TYPE | Release |
| JENKINS_COMM_TYPE | MPI |
| JENKINS_DO_COMPLEX | OFF |
| JENKINS_JOB_TYPE | Experimental |
| MPI_MODULE | sems-openmpi/1.8.7 |
| TEST_REPO_ALIAS | TRILINOS |
| TRILINOS_SOURCE_BRANCH | KKWork |
| TRILINOS_SOURCE_REPO | https://github.com/brian-kelley/Trilinos |
| TRILINOS_SOURCE_SHA | 21c2370124f9f2a84b544b120d632dc59b5a68f2 |
| TRILINOS_TARGET_BRANCH | develop |
| TRILINOS_TARGET_REPO | https://github.com/trilinos/Trilinos |
| TRILINOS_TARGET_SHA | 4b51a54ec727088dd3666053a4d61b5d1d0ac6ec |

## Build Information
- Test Name: Trilinos_pullrequest_gcc_4.9.3
- Build Num: 42
- Status: STARTED

### Jenkins Parameters
| Parameter Name | Value |
|----------------|-------|
| COMPILER_MODULE | sems-gcc/4.9.3 |
| JENKINS_BUILD_TYPE | Release |
| JENKINS_COMM_TYPE | MPI |
| JENKINS_DO_COMPLEX | OFF |
| JENKINS_JOB_TYPE | Experimental |
| MPI_MODULE | sems-openmpi/1.8.7 |
| TEST_REPO_ALIAS | TRILINOS |
| TRILINOS_SOURCE_BRANCH | KKWork |
| TRILINOS_SOURCE_REPO | https://github.com/brian-kelley/Trilinos |
| TRILINOS_SOURCE_SHA | 21c2370124f9f2a84b544b120d632dc59b5a68f2 |
| TRILINOS_TARGET_BRANCH | develop |
| TRILINOS_TARGET_REPO | https://github.com/trilinos/Trilinos |
| TRILINOS_TARGET_SHA | 4b51a54ec727088dd3666053a4d61b5d1d0ac6ec |

## Using Repos:
- Repo: TRILINOS (brian-kelley/Trilinos)
- Branch: KKWork
- SHA: 21c2370124f9f2a84b544b120d632dc59b5a68f2
- Mode: TEST_REPO

Pull Request Author: brian-kelley

trilinos-autotester commented:
Status Flag 'Pull Request AutoTester' - Jenkins Testing: all Jobs PASSED

Pull Request Auto Testing has PASSED
------------------------------------

## Build Information
- Test Name: Trilinos_autotester_test
- Build Num: 125
- Status: PASSED

### Jenkins Parameters
| Parameter Name | Value |
|----------------|-------|
| COMPILER_MODULE | sems-gcc/4.8.4 |
| JENKINS_BUILD_TYPE | Release |
| JENKINS_COMM_TYPE | MPI |
| JENKINS_DO_COMPLEX | OFF |
| JENKINS_JOB_TYPE | Experimental |
| MPI_MODULE | sems-openmpi/1.8.7 |
| TEST_REPO_ALIAS | TRILINOS |
| TRILINOS_SOURCE_BRANCH | KKWork |
| TRILINOS_SOURCE_REPO | https://github.com/brian-kelley/Trilinos |
| TRILINOS_SOURCE_SHA | 21c2370124f9f2a84b544b120d632dc59b5a68f2 |
| TRILINOS_TARGET_BRANCH | develop |
| TRILINOS_TARGET_REPO | https://github.com/trilinos/Trilinos |
| TRILINOS_TARGET_SHA | 4b51a54ec727088dd3666053a4d61b5d1d0ac6ec |

## Build Information
- Test Name: Trilinos_pullrequest_gcc_4.9.3
- Build Num: 42
- Status: PASSED

### Jenkins Parameters
| Parameter Name | Value |
|----------------|-------|
| COMPILER_MODULE | sems-gcc/4.9.3 |
| JENKINS_BUILD_TYPE | Release |
| JENKINS_COMM_TYPE | MPI |
| JENKINS_DO_COMPLEX | OFF |
| JENKINS_JOB_TYPE | Experimental |
| MPI_MODULE | sems-openmpi/1.8.7 |
| TEST_REPO_ALIAS | TRILINOS |
| TRILINOS_SOURCE_BRANCH | KKWork |
| TRILINOS_SOURCE_REPO | https://github.com/brian-kelley/Trilinos |
| TRILINOS_SOURCE_SHA | 21c2370124f9f2a84b544b120d632dc59b5a68f2 |
| TRILINOS_TARGET_BRANCH | develop |
| TRILINOS_TARGET_REPO | https://github.com/trilinos/Trilinos |
| TRILINOS_TARGET_SHA | 4b51a54ec727088dd3666053a4d61b5d1d0ac6ec |

trilinos-autotester commented:
Status Flag 'Pre-Merge Inspection' - - This Pull Request Requires Inspection... The code must be inspected by a member of the Team before Testing/Merging
           THE LAST COMMIT TO THIS PULL REQUEST HAS NOT BEEN REVIEWED YET!

trilinos-autotester commented:
All Jobs Finished; status = PASSED, However Inspection must be performed before merge can occur...

trilinos-autotester commented:
All Jobs Finished; status = PASSED, However Inspection must be performed before merge can occur...

trilinos-autotester commented:
All Jobs Finished; status = PASSED, However Inspection must be performed before merge can occur...

trilinos-autotester commented:
All Jobs Finished; status = PASSED, However Inspection must be performed before merge can occur...

trilinos-autotester commented:
All Jobs Finished; status = PASSED, However Inspection must be performed before merge can occur...

trilinos-autotester commented:
All Jobs Finished; status = PASSED, However Inspection must be performed before merge can occur...

trilinos-autotester commented:
All Jobs Finished; status = PASSED, However Inspection must be performed before merge can occur...

trilinos-autotester commented:
All Jobs Finished; status = PASSED, However Inspection must be performed before merge can occur...

trilinos-autotester commented:
All Jobs Finished; status = PASSED, However Inspection must be performed before merge can occur...

trilinos-autotester commented:
All Jobs Finished; status = PASSED, However Inspection must be performed before merge can occur...

trilinos-autotester commented:
All Jobs Finished; status = PASSED, However Inspection must be performed before merge can occur...

trilinos-autotester commented:
All Jobs Finished; status = PASSED, However Inspection must be performed before merge can occur...

trilinos-autotester commented:
All Jobs Finished; status = PASSED, However Inspection must be performed before merge can occur...

trilinos-autotester commented:
All Jobs Finished; status = PASSED, However Inspection must be performed before merge can occur...

trilinos-autotester commented:
All Jobs Finished; status = PASSED, However Inspection must be performed before merge can occur...

trilinos-autotester commented:
All Jobs Finished; status = PASSED, However Inspection must be performed before merge can occur...

trilinos-autotester commented:
All Jobs Finished; status = PASSED, However Inspection must be performed before merge can occur...

trilinos-autotester commented:
All Jobs Finished; status = PASSED, However Inspection must be performed before merge can occur...

trilinos-autotester commented:
All Jobs Finished; status = PASSED, However Inspection must be performed before merge can occur...

trilinos-autotester commented:
All Jobs Finished; status = PASSED, However Inspection must be performed before merge can occur...

trilinos-autotester commented:
All Jobs Finished; status = PASSED, However Inspection must be performed before merge can occur...

trilinos-autotester commented:
All Jobs Finished; status = PASSED, However Inspection must be performed before merge can occur...

trilinos-autotester commented:
All Jobs Finished; status = PASSED, However Inspection must be performed before merge can occur...

trilinos-autotester commented:
All Jobs Finished; status = PASSED, However Inspection must be performed before merge can occur...

trilinos-autotester commented:
All Jobs Finished; status = PASSED, However Inspection must be performed before merge can occur...

trilinos-autotester commented:
All Jobs Finished; status = PASSED, However Inspection must be performed before merge can occur...

trilinos-autotester commented:
All Jobs Finished; status = PASSED, However Inspection must be performed before merge can occur...

trilinos-autotester commented:
All Jobs Finished; status = PASSED, However Inspection must be performed before merge can occur...

trilinos-autotester commented:
All Jobs Finished; status = PASSED, However Inspection must be performed before merge can occur...

trilinos-autotester commented:
All Jobs Finished; status = PASSED, However Inspection must be performed before merge can occur...

trilinos-autotester commented:
All Jobs Finished; status = PASSED, However Inspection must be performed before merge can occur...

ibaned commented:
Marking this `[WIP]` so the autotester will ignore it until it gets some attention again. Simply remove `[WIP]` from the title to restart autotesting.

brian-kelley commented:
Deleting this PR because the KokkosKernels snapshot with the addition kernels is now in Trilinos. Will put in a different PR to use KokkosKernels addition in Tpetra and remove the (now redundant) kernels now in TpetraExt_MatrixMatrix_def.hpp.
